source/extensions/quic_listeners/quiche/platform/quic_mem_slice_impl.h

```diff
@@ -0,0 +1,87 @@
+#pragma once
+
+// NOLINT(namespace-envoy)
+
+// This file is part of the QUICHE platform implementation, and is not to be
+// consumed or referenced directly by other Envoy code. It serves purely as a
+// porting layer for QUICHE.
+
+#include <iostream>
+#include <memory>
+
+#include "common/buffer/buffer_impl.h"
+#include "common/common/assert.h"
+
+#include "quiche/quic/core/quic_buffer_allocator.h"
+
+namespace quic {
+
+class QuicMemSliceImpl {
+public:
+  // Constructs an empty QuicMemSliceImpl.
+  QuicMemSliceImpl() = default;
+
+  // Constructs a QuicMemSliceImpl by let |allocator| allocate a data buffer of
+  // |length|.
+  QuicMemSliceImpl(QuicBufferAllocator* allocator, size_t length)
+      : length_(length), slice_(new Envoy::Buffer::OwnedImpl()) {
+    auto fragment = new Envoy::Buffer::BufferFragmentImpl(
+        allocator->New(length), length,
+        [allocator](const void* data, size_t, const Envoy::Buffer::BufferFragmentImpl* fragment) {
+          allocator->Delete(const_cast<char*>(static_cast<const char*>(data)));
+          delete fragment;
+        });
+    slice_->addBufferFragment(*fragment);
+  }
+
+  // Constructs a QuicMemSliceImpl from a Buffer::Instance whose getRawSlices()
+  // returns only 1 slice.
+  QuicMemSliceImpl(std::shared_ptr<Envoy::Buffer::Instance> slice)
+      : length_(slice->length()), slice_(std::move(slice)) {
+    ASSERT(slice_->getRawSlices(nullptr, 0) == 1);
+  }
+
+  QuicMemSliceImpl(const QuicMemSliceImpl& other) = delete;
+  QuicMemSliceImpl& operator=(const QuicMemSliceImpl& other) = delete;
+
+  // Move constructors. |other| will not hold a reference to the data buffer
+  // after this call completes.
+  QuicMemSliceImpl(QuicMemSliceImpl&& other)
+      : length_(other.length_), slice_(std::move(other.slice_)) {
+    other.Reset();
+  }
+
+  QuicMemSliceImpl& operator=(QuicMemSliceImpl&& other) {
+    if (this != &other) {
+      length_ = other.length_;
+      slice_ = std::move(other.slice_);
+      other.Reset();
+    }
+    return *this;
+  }
+
+  void Reset() {
```

Comment:
As this is in the Envoy tree, can we use in this class the Envoy convention that when you implement an interface to immediately precede all such implementations with a comment saying whose interface you are implementing? E.g.
```
  // QuicMemSlice
  void Reset() {...}
  const char* data() { ... }
```
I realize you are not actually implementing a virtual interface but I still think it's helpful. In lots of classes in Envoy it's very useful to see that because there is a ton of multiple inheritance and we need to know which superclass's interface is being implemented. Here, it's especially useful because the methods are not formally c++ `override` it will tell us which interfaces are required for the plug-in, and which are helpers or extra methods provided to end-users.